walrus_mainboard_driver/src/walrus_mainboard_driver/walrus_mainboard_driver.cpp

```diff
@@ -0,0 +1,670 @@
+                    msg2.msg = "";
+                    to_board.publish(msg2);
+                }
+                batteries[l].present = true;
+                batteries[l].upper_voltage = msg.lcell_voltage[l] / 1000.0;
+                batteries[l].upper_current = msg.lcell_current[l] / 1000.0;
+                batteries[l].upper_charge = msg.lcell_charge[l] / 100.0;
+                batteries[l].upper_temp = msg.lcell_temp[l] / 100.0;
+                batteries[l].upper_avg_current = msg.lcell_avgcurr[l] / 1000.0;
+                batteries[l].upper_voltage = msg.ucell_voltage[l] / 1000.0;
+                batteries[l].upper_current = msg.ucell_current[l] / 1000.0;
+                batteries[l].upper_charge = msg.ucell_charge[l] / 100.0;
+                batteries[l].upper_temp = msg.ucell_temp[l] / 100.0;
+                batteries[l].upper_avg_current = msg.ucell_avgcurr[l] / 1000.0;
+                batteries[l].shutdown = (batteries[l].lower_voltage < 2 || batteries[l].upper_voltage < 2);
+                batteries[l].combined_voltage = batteries[l].lower_voltage + batteries[l].upper_voltage;
+                batteries[l].combined_current = (batteries[l].lower_current + batteries[l].upper_current) / 2;
+                batteries[l].combined_avg_current = (batteries[l].lower_avg_current + batteries[l].upper_avg_current) / 2;
+                batteries[l].combined_charge = (batteries[l].lower_charge + batteries[l].upper_charge) / 2;
+                main_voltage += batteries[l].combined_voltage;
+                main_current += batteries[l].combined_current;
+                main_avg_current += batteries[l].combined_avg_current;
+                main_charge += batteries[l].combined_charge;
+                count++;
+            }
+            else
+                batteries[l].present = false;
+        }
+        main_voltage /= count;
+        main_current /= count;
+        main_avg_current /= count;
+        main_charge /= count;
+    }
+    
+	void MainBoardDriver::from_board_callback(const walrus_firmware_msgs::MainBoardControl& msg)
+	{
+	    boost::lock_guard<boost::mutex> lock(sensor_data_mutex);
+	    
+	    switch (msg.type)
+	    {
+            case walrus_firmware_msgs::MainBoardControl::BATT_MFR:
+                batteries[msg.index].mfr = msg.msg;
+            break;
+            case walrus_firmware_msgs::MainBoardControl::BATT_NAME:
+                batteries[msg.index].dev_name = msg.msg;
+            break;
+            case walrus_firmware_msgs::MainBoardControl::BATT_CHEM:
+                batteries[msg.index].chemistry = msg.msg;
+            break; 
+            case walrus_firmware_msgs::MainBoardControl::STATUS:
+                main_board_status = msg.msg;
+                main_board_status_level = msg.value;
+            break;
+	    }
+    }
+    
+    //Diagnostics updaters
+    void MainBoardDriver::chassis_diagnostic_callback(diagnostic_updater::DiagnosticStatusWrapper &stat)
```

Comment:
You could break this up so that each sensor is it's own status item and then just group them in the aggregator, including a separate one for each water sensor